Azure.Compute.Batch inital checkin (#44524)
Azure/azure-sdk-for-net: sdk/batch/Azure.Compute.Batch/samples/Sample1_CreatePool_Job_Task.md

# Creating a Batch Pool, Job, and Tasks

This sample demonstrates how to create a pool, job, and tasks for a Batch account.

> ## Special Note for Batch Pools

Reviewer: ```suggestion
## Special note for Batch Pools
```

Both the **Azure.Compute.Batch** and the ARM based **Azure.Management.Batch** sdk's support the creating and managing of Batch Pools with the only difference being that **Azure.Management.Batch** supports creating [Batch Pools with Managed Idenities](https://learn.microsoft.com/azure/batch/managed-identity-pools), for this reason creating pools via **Azure.Management.Batch** is considered a best practice.  This example will use **Azure.Management.Batch** for pool creation and thus will need to create an ArmClient to do so.

## Authenticating the Azure.ResourceManager `ArmClient`

In order to create Batch Pool from the Azure.Managment.Batch library you will need to instantiate an Armclient. To create an authenticated client and start interacting with Microsoft Azure resources, see the [quickstart guide here](https://github.com/Azure/azure-sdk-for-net/blob/main/doc/dev/mgmt_quickstart.md).

> ```C#

Reviewer: As mentioned in the README, these need to be snippets.

Author: done

var credential = new DefaultAzureCredential();
ArmClient _armClient = new ArmClient(credential);
```

> 
> 

Reviewer: ```suggestion

```

### Pool Creation

Batch operations in the **Azure.Management.Batch** sdk are preformed from a BatchAccountResource object, to get a BatchAccountResource object you can query the armclient for the resource id of your Batch account.
```C#
BatchAccountResource batchAccount = await _armClient.GetBatchAccountResource("your-batch-account-resource-id").GetAsync();
```

With the BatchAccountResource you can create a pool with the [batchAccount.GetBatchAccountPools().CreateOrUpdateAsync](https://learn.microsoft.com/dotnet/api/azure.resourcemanager.batch.batchaccountpoolcollection.createorupdateasync?view=azure-dotnet) command

> ```C#

Reviewer: ```suggestion

```C#
```

> 

Reviewer: ```suggestion
```

var poolName = "HelloWorldPool";
var imageReference = new BatchImageReference()
{
    Publisher = "canonical",
    Offer = "0001-com-ubuntu-server-jammy",
    Sku = "22_04-lts",
    Version = "latest"
};
string nodeAgentSku = "batch.node.ubuntu 22.04";

BatchAccountPoolResource pool = (await batchAccount.GetBatchAccountPools().CreateOrUpdateAsync(WaitUntil.Completed, poolName, new BatchAccountPoolData()
{
    VmSize = "Standard_DS1_v2",
    DeploymentConfiguration = new BatchDeploymentConfiguration()
    {
        VmConfiguration = new BatchVmConfiguration(imageReference, nodeAgentSku)
    },
    ScaleSettings = new BatchAccountPoolScaleSettings()
    {
        FixedScale = new BatchAccountFixedScaleSettings()
        {
            TargetDedicatedNodes = 1
        }
    }
})).Value;
```

> ## Authenticating the Azure.Compute.Batch `BatchClient`

Reviewer: ```suggestion
## Authenticating the Azure.Compute.Batch `BatchClient`

```

Creation of Batch jobs and tasks can only be preformed with the `Azure.Compute.Batch` sdk.  A `BatchClient` object is needed to preform Batch operations and can be created using [Microsoft Entra ID authtentication](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity/README.md) and the Batch account endpoint  

```C#
var credential = new DefaultAzureCredential();
BatchClient _batchClient = new BatchClient(
    new Uri("https://examplebatchaccount.eastus.batch.azure.com"),
    credential);
```

> ### Job Creation

Reviewer: ```suggestion
### Job creation
```

nit:  The Microsoft style guide requires that we use sentence-style capitalization in titles and headings.   _([source](https://learn.microsoft.com/style-guide/capitalization))_

Before we can create Batch Tasks we first need to create a Job for the tasks to be associatd with, this can be done via the `CreateJobAsync` command. The basic elements needed are a Id for job itself and the name of the pool that this job will run against. 

> ```C#

Reviewer: ```suggestion

```C#
```

await _batchClient.CreateJobAsync(new BatchJobCreateContent(jobId, new BatchPoolInfo() { PoolId = poolName }));
```

> ### Task Creation

Reviewer: ```suggestion

### Task Creation

```

Batch tasks can be created from the BatchClient via the `CreateTaskAsync`.  The basic elements needed are the name of the job the task will be assigned to, and id for the task itself, and a command to run.

> ```C#

Reviewer: ```suggestion

```C#
```

await _batchClient.CreateTaskAsync(jobId, new BatchTaskCreateContent(taskId, $"echo Hello world from {taskId}"));
```

> ### Full Sample

Reviewer: ```suggestion

### Full Sample
```

Reviewer: This is not something that we do.   If we have a full end-to-end sample, we'll want to package it as such in a way that builds in and out of this repository and which can be exported to the Azure samples browser.

- [Example: Key Vault](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/keyvault/samples)
- [Example: Service Bus](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/servicebus/Azure.Messaging.ServiceBus/samples/DeadLetterQueue)

Author: removed.

# Creating a Batch Pool, Job, and Tasks

This sample demonstrates how to create a pool, job, and tasks for a Batch account.

## Special note for Batch pools

> Both the **Azure.Compute.Batch** and the ARM based **Azure.Management.Batch** sdk's support the creating and managing of Batch Pools with the only difference being that **Azure.Management.Batch** supports creating [Batch Pools with Managed Idenities](https://learn.microsoft.com/azure/batch/managed-identity-pools), for this reason creating pools via **Azure.Management.Batch** is considered a best practice.  This example will use **Azure.Management.Batch** for pool creation and thus will need to create an ArmClient to do so.

Reviewer: ```suggestion
Both the **Azure.Compute.Batch** and the ARM-based **Azure.ResourceManager.Batch** libraries support the creating and managing of Batch Pools with the only difference being that **Azure.Management.Batch** supports creating [Batch Pools with Managed Identities](https://learn.microsoft.com/azure/batch/managed-identity-pools), for this reason creating pools via **Azure.ResourceManager.Batch** is considered a best practice.  This example will use **Azure.ResourceManager.Batch** for pool creation and thus will need to create an `ArmClient` to do so.
```

nit: branding.   "Azure SDK" is the all-up product which contains a number of client and management libraries.  We've been asked not to use SDK to refer to an individual package.

## Authenticating the Azure.ResourceManager `ArmClient`

> In order to create Batch Pool from the Azure.Managment.Batch library you will need to instantiate an Armclient. To create an authenticated client and start interacting with Microsoft Azure resources, see the [quickstart guide here](https://github.com/Azure/azure-sdk-for-net/blob/main/doc/dev/mgmt_quickstart.md).

Reviewer: All of the references to the management library need a search and replace.   The current generation package is `Azure.ResourceManager.Batch`.    

_(the previous generation, which reached end of life and is no longer supported, was `Microsoft.Azure.Management.Batch`)_

Author: fixed

> In order to create Batch Pool from the Azure.Managment.Batch library you will need to instantiate an Armclient. To create an authenticated client and start interacting with Microsoft Azure resources, see the [quickstart guide here](https://github.com/Azure/azure-sdk-for-net/blob/main/doc/dev/mgmt_quickstart.md).

Reviewer: ```suggestion
In order to create a Batch Pool from the Azure.Managment.Batch library you will need to instantiate an Armclient. To create an authenticated client and start interacting with Microsoft Azure resources, see the [quickstart guide](https://github.com/Azure/azure-sdk-for-net/blob/main/doc/dev/mgmt_quickstart.md).
```

```C# Snippet:Batch_Sample01_CreateBatchMgmtClient
var credential = new DefaultAzureCredential();
ArmClient _armClient = new ArmClient(credential);
```

> ### Pool creation

Reviewer: ```suggestion

### Pool creation
```

Batch operations in the **Azure.Management.Batch** sdk are preformed from a BatchAccountResource object, to get a BatchAccountResource object you can query the armclient for the resource id of your Batch account.

> ```C# Snippet:Batch_Sample01_GetBatchMgmtAccount

Reviewer: ```suggestion

```C# Snippet:Batch_Sample01_GetBatchMgmtAccount
```

var batchAccountIdentifier = ResourceIdentifier.Parse("your-batch-account-resource-id");
BatchAccountResource batchAccount = await _armClient.GetBatchAccountResource(batchAccountIdentifier).GetAsync();
```

With the BatchAccountResource you can create a pool with the [batchAccount.GetBatchAccountPools().CreateOrUpdateAsync](https://learn.microsoft.com/dotnet/api/azure.resourcemanager.batch.batchaccountpoolcollection.createorupdateasync?view=azure-dotnet) command

> ```C# Snippet:Batch_Sample01_PoolCreation

Reviewer: ```suggestion

```C# Snippet:Batch_Sample01_PoolCreation
```

var poolName = "HelloWorldPool";
var imageReference = new BatchImageReference()
{
    Publisher = "canonical",
    Offer = "0001-com-ubuntu-server-jammy",
    Sku = "22_04-lts",
    Version = "latest"
};
string nodeAgentSku = "batch.node.ubuntu 22.04";

> BatchAccountPoolResource pool = (await batchAccount.GetBatchAccountPools().CreateOrUpdateAsync(WaitUntil.Completed, poolName, new BatchAccountPoolData()

Reviewer: I'd suggest breaking this into two lines.   The heavily nested parenthesis and skipping of the operation return type to get to `Value` is going to be confusing for some.

Author: broke into two lines

{
    VmSize = "Standard_DS1_v2",
    DeploymentConfiguration = new BatchDeploymentConfiguration()
    {
        VmConfiguration = new BatchVmConfiguration(imageReference, nodeAgentSku)
    },
    ScaleSettings = new BatchAccountPoolScaleSettings()
    {
        FixedScale = new BatchAccountFixedScaleSettings()
        {
            TargetDedicatedNodes = 1
        }
    }
})).Value;
```

## Authenticating the Azure.Compute.Batch `BatchClient`

> Creation of Batch jobs and tasks can only be preformed with the `Azure.Compute.Batch` sdk.  A `BatchClient` object is needed to preform Batch operations and can be created using [Microsoft Entra ID authtentication](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity/README.md) and the Batch account endpoint  

Reviewer: ```suggestion
Creation of Batch jobs and tasks can only be preformed with the `Azure.Compute.Batch` library.  A `BatchClient` object is needed to preform Batch operations and can be created using [Microsoft Entra ID authtentication](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity/README.md) and the Batch account endpoint  
```

also acceptable would be "package."    We should do a scrub and make sure that there are no references to "sdk" left in here.

> Creation of Batch jobs and tasks can only be preformed with the `Azure.Compute.Batch` sdk.  A `BatchClient` object is needed to preform Batch operations and can be created using [Microsoft Entra ID authtentication](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity/README.md) and the Batch account endpoint  

Reviewer: ```suggestion
Creation of Batch jobs and tasks can only be preformed with the `Azure.Compute.Batch` library.  A `BatchClient` instance is needed to preform Batch operations and can be created using [Microsoft Entra ID authtentication](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity/README.md) and the Batch account endpoint  
```

Author: updated

```C# Snippet:Batch_Sample01_CreateBatchClient
var credential = new DefaultAzureCredential();
BatchClient _batchClient = new BatchClient(
new Uri("https://examplebatchaccount.eastus.batch.azure.com"), credential);
```

### Job creation

> Before we can create Batch Tasks we first need to create a Job for the tasks to be associatd with, this can be done via the `CreateJobAsync` command. The basic elements needed are a Id for job itself and the name of the pool that this job will run against. 

Reviewer: ```suggestion
Before we can create Batch Tasks, we first need to create a Job for the tasks to be associatd with, this can be done via the `CreateJobAsync` command. The basic elements needed are an id for job itself and the name of the pool that this job will run against. 
```

also acceptable "ID"

> ```C# Snippet:Batch_Sample01_CreateBatchJob

Reviewer: ```suggestion

```C# Snippet:Batch_Sample01_CreateBatchJob
```

await _batchClient.CreateJobAsync(new BatchJobCreateContent("jobId", new BatchPoolInfo() { PoolId = "poolName" }));
```

### Task creation

Batch tasks can be created from the BatchClient via the `CreateTaskAsync`.  The basic elements needed are the name of the job the task will be assigned to, and id for the task itself, and a command to run.
```C# Snippet:Batch_Sample01_CreateBatchTask
await _batchClient.CreateTaskAsync("jobId", new BatchTaskCreateContent("taskId", $"echo Hello world"));
```

> ### Task results

Reviewer: ```suggestion

### Task results
```

Onces the tasks are complete `GetTasksAsync` cand be used to retrieve the `BatchTask` object and `GetTaskFileAsync` can be used to get the output files

```C# Snippet:Batch_Sample01_GetTasks

> var completedTasks = _batchClient.GetTasksAsync(jobId, filter: "state eq 'completed'");

Reviewer: What is `_batchClient` and how do I make one?   Snippets are intended to be fully copy/paste and should be entirely self-contained and complete.   

Developers shouldn't have to go elsewhere and find other boilerplate to make a snippet work - just paste into their app and compile.

Author: I was kind of following the keyvalut examples where they state that all the samples will use the client created above https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/keyvault/Azure.Security.KeyVault.Secrets#examples
